### PR TITLE
Optional SSH Support for DSR Processing

### LINKED ISSUE #3329: Provide Optional SSH Tunneling for Database Connections
### Is your feature request related to a specific problem?

A common request today is around more choices in securing database connections. One option that will provide some better flexibility would be to allow the usage of an SSH tunnel.

### Describe the solution you'd like

Support SSH tunnels for database connectors, possibly as an optional item when setting up the connection config

### Describe alternatives you've considered, if any

Some of the work around distributed fides (#1861) could also be applied here

### Additional context

This will likely require quite a bit of manual testing and some help from @RobertKeyser to validate

### PR DESCRIPTION
**Pull request by SteveDMurphy**
Closes #3329

### Code Changes

* [x] Create the config variables for ssh config
* [x] Create a boolean to require ssh on specific connectors
* [x] Functionalize the creation of an SSH tunnel as required
* [x] Update the query config to use the SSH tunnel as required
* [x] ~Look at the prospect of having the `generate` and `classify` commands use the SSH tunnel in this first iteration~
* [x] Add SSH URI for both Postgres and RedShift

### Steps to Confirm

* [ ] Set the environment variables as defined in [this 1Password secure note](https://start.1password.com/open/i?a=5LPILPRDUBGWBPJWYS6QHHHLRE&v=p3bpsusghc3t45pzpsebehrd4m&i=ofqbbrdfdzadtizxg33azulzce&h=ethyca.1password.com) - (these will all be exchanged with some proper test credentials)
* [ ] If on an M1, add [these commands](https://github.com/ethyca/fides/pull/3374/commits/10b1b728c10d56fd22caf9514af7a9d438be5c7b#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R68-R72) to the Dockerfile to handle a known issue
* [ ] `nox -s dev -- shell`
* [ ] Create a fake dataset: `fides user login && fides push`
* [ ] Connect to the db behind the vpc via the bastion host with [these credentials](https://start.1password.com/open/i?a=5LPILPRDUBGWBPJWYS6QHHHLRE&v=p3bpsusghc3t45pzpsebehrd4m&i=ofqbbrdfdzadtizxg33azulzce&h=ethyca.1password.com), connecting to your pushed dataset
* [ ] Validate the connection test succeeds

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### Description Of Changes

This is currently set up to only support a single bastion host in order to get moving more quickly. In a follow on change, we will likely create an SSH connection config type to support multiple ssh requirements.

### CODE REVIEW

File: src/fides/api/schemas/connection_configuration/connection_secrets_postgres.py [hunk 1]
```diff
@@ -17,6 +17,7 @@ class PostgreSQLSchema(ConnectionConfigSecretsSchema):
         str
     ] = None  # Either the entire "url" *OR* the "host" should be supplied.
     port: Optional[int] = None
+    ssh_required: bool = False
```
Comment (SteveDMurphy) [inline src/fides/api/schemas/connection_configuration/connection_secrets_postgres.py:20]: Is there a better name or description to be applied here?

**Review by SteveDMurphy — COMMENTED** [review]
Still need to add test infrastructure, have it added to secrets, then write tests and docs

File: Dockerfile [hunk 2]
```diff
@@ -64,6 +64,8 @@ RUN if [ "$TARGETPLATFORM" != "linux/arm64" ] ; \
     && rm -rf /var/lib/apt/lists/* ; \
     fi
 
+
+
```
Comment (seanpreston) [inline Dockerfile:68]: Is this erroneous?

File: docker-compose.yml [hunk 3]
```diff
@@ -28,6 +28,9 @@ services:
       FIDES__DEV_MODE: "True"
       FIDES__USER__ANALYTICS_OPT_OUT: "True"
       FIDES__SECURITY__ALLOW_CUSTOM_CONNECTOR_FUNCTIONS: "True"
+      FIDES__SECURITY__BASTION_SERVER_HOST: ${FIDES__SECURITY__BASTION_SERVER_HOST-}
+      FIDES__SECURITY__BASTION_SERVER_SSH_USERNAME: ${FIDES__SECURITY__BASTION_SERVER_SSH_USERNAME-}
+      FIDES__SECURITY__BASTION_SERVER_SSH_PKEY: ${FIDES__SECURITY__BASTION_SERVER_SSH_PKEY-}
```
Comment (seanpreston) [inline docker-compose.yml:33]: I know it's awful long already but would be good to be explicit about `PKEY` — public vs. private has big implications here :)

File: src/fides/api/schemas/connection_configuration/connection_secrets_postgres.py (hunk 1 above)
Comment (seanpreston) [reply src/fides/api/schemas/connection_configuration/connection_secrets_postgres.py:20]: This is great

File: src/fides/api/service/connectors/sql_connector.py [hunk 4]
```diff
@@ -176,6 +185,24 @@ def set_schema(self, connection: Connection) -> None:
         """Optionally override to set the schema for a given database that
         persists through the entire session"""
 
+    def create_ssh_tunnel(self, host: Optional[str], port: Optional[int]) -> None:
+        """Creates an SSH Tunnel to forward ports as configured."""
+        if CONFIG.security.bastion_server_ssh_pkey:
```
Comment (seanpreston) [inline src/fides/api/service/connectors/sql_connector.py:190]: Should we raise a custom exception here if this is not set, but `ssh_required` has led Fides down this path?

Comment (SteveDMurphy) [reply src/fides/api/service/connectors/sql_connector.py:190]: That seems like a great idea, thanks!

File: docker-compose.yml (hunk 3 above)
Comment (SteveDMurphy) [reply docker-compose.yml]: thanks for fixing!

**Review by seanpreston — APPROVED** [review]
LGTM @SteveDMurphy — would be good to get a follow-up ticket in for some more thorough testing

**Comment by SteveDMurphy** [comment]
> LGTM @SteveDMurphy — would be good to get a follow-up ticket in for some more thorough testing

Manual testing looks good and added #3562 for the testing follow on - thanks @seanpreston and @RobertKeyser !!

Will merge once checks complete

**Pull request merged**